src/vm/amd64/CrtHelpers.asm

```diff
@@ -13,516 +13,264 @@
-        ret                             ;
-
-;
-; Move residual bytes.
-;
-
-        align
-
-mcpy50: mov     al, [rcx + rdx]         ; move byte
-        mov     [rcx], al               ;
-        inc     rcx                     ; increment destiantion address
-        dec     r8                      ; decrement loop count
-        jnz     short mcpy50            ; if nz, more bytes to fill
-        mov     rax, r11                ; set destination address
-        ret                             ; return
-
-;
-; Move 32 byte blocks
-;
-
-        align   16
-
-        db      066h, 066h, 066h, 090h
-        db      066h, 066h, 090h
+        mov     rax, rcx                ; save dst address
+        cmp     r8, 16                  
+        jbe     mcpy02
+        
+        cmp     r8, 64             
+        jnbe    mcpy07
+
+        ; count > 16 && count <= 64
+        align 16        
+mcpy00: movdqu  xmm0, [rdx]             
+        movdqu  xmm1, -16[rdx + r8]     ; save 16 to 32 bytes src
+        cmp     r8, 32
+        jbe     mcpy01
+        
+        movdqu  xmm2, 16[rdx]           
+        movdqu  xmm3, -32[rdx + r8]     ; save 32 to 64 bytes src
+        
+        ;count > 32 && count <= 64
+        movdqu  16[rcx], xmm2
+        movdqu  -32[rcx + r8], xmm3
+        
+        ;count > 16 && count <= 32
+mcpy01: movdqu  [rcx], xmm0
+        movdqu  -16[rcx + r8], xmm1
+        ret
+
+        ; count <= 16 
+        align 16
+mcpy02: test    r8b, 24                 ; test count with 0001 1000
+        jz      mcpy03
+        ; count >= 8 && count <= 16
+        mov     r9, [rdx]
+        mov     r10, -8[rdx + r8]
+        mov     [rcx], r9
+        mov     -8[rcx + r8], r10
+        ret
+        
+        align 16
+mcpy03: test    r8b, 4                  ; test count with 0100
+        jz      mcpy04
+        ; count >= 4 && count < 8
+        mov     r9d, [rdx]
+        mov     r10d, -4[rdx + r8]
+        mov     [rcx], r9d
+        mov     -4[rcx + r8], r10d
+        ret
+        
+        ; count >= 0 && count < 4
+        align 16
+mcpy04: test    r8, r8                  
+        jz      mcpy06                  ; count == 1/2/3
+        mov     r9b, [rdx]              ; save the first byte
+        
+        test    r8b, 2                  ; test count with 0010
+        jz      mcpy05
+        mov     r10w, -2[rdx + r8]        
+        mov     -2[rcx + r8], r10w
+mcpy05: mov     [rcx], r9b
+mcpy06: ret
+ 
+        align 16
+        ; count > 64, we need to check overlap
+mcpy07: mov     r9, rdx                 ; r9 is src address
+        sub     r9, rcx                 ; if src - dst < 0 jump to mcpy11
+        jb      mcpy11                  ; if b, destination may overlap 
+        
+mcpy08: cmp     r8, 512
+        jnbe    mcpy10
+        
+        ; count > 64 && count <= 512
+        mov     r9, r8
+        shr     r9, 6                   ; count/64
+        
+        align 16
+mcpy09: movdqu  xmm0, [rdx] 
```

Comment:
For certain copy lengths, we do see improvement when using YMM with vzeroupper. However, considering the cpu dispatch cost, we may consider YMM as future improvement.